[bug] Libraries can contain characters that are not allowed in cmake target names

### Environment Details (include every applicable attribute)
  * Operating System+version: Linux 5.12.11-arch1-1 x86_64
  * Conan version: 1.36.0
  * Python version: 3.9.5
  * CMake version: 3.20.4

### Steps to reproduce (Include if Applicable)
Install a conan package where one of its libraries has an `@` in the name with the `cmake_find_package` generator.
When calling `find_package()` for this package CMake will fail due to the `@`.

I believe this is the relevant docs from CMake: https://cmake.org/cmake/help/latest/policy/CMP0037.html
I might be able to work around it by removing those libs from `cpp_info.libs` and add a flag to `cpp_info.sharedlinkflags`.

### Logs (Executed commands with output) (Include/Attach if Applicable)
In my case I have created a conan package that exports the [Android VNDK](https://source.android.com/devices/architecture/vndk) from a build of AOSP. They have many, many libraries of the format: `<library>@<version>.so`.

CMake gives this error:
```
CMake Error at build_vndk/Findandroid-vndk.cmake:39 (add_library):
  The target name "CONAN_LIB::android-vndk_stagefright_bufferpool@2.0" is
  reserved or not valid for certain CMake features, such as generator
  expressions, and may result in undefined behavior.
Call Stack (most recent call first):
  build_vndk/Findandroid-vndk.cmake:125 (conan_package_library_targets)
  src/CMakeLists.txt:7 (find_package)


CMake Error at build_vndk/Findandroid-vndk.cmake:40 (set_target_properties):
  set_target_properties Can not find target to add properties to:
  CONAN_LIB::android-vndk_stagefright_bufferpool@2.0
Call Stack (most recent call first):
  build_vndk/Findandroid-vndk.cmake:125 (conan_package_library_targets)
  src/CMakeLists.txt:7 (find_package)
```

--- a/conan/tools/cmake/cmakedeps/templates/macros.py
+++ b/conan/tools/cmake/cmakedeps/templates/macros.py
@@ -58,6 +58,8 @@ class MacrosTemplate(CMakeDepsFileTemplate):
                    list(APPEND _out_libraries ${CONAN_FOUND_LIBRARY})
 
                    # Create a micro-target for each lib/a found
+                   # Allow only some characters for the target name
+                   string(REGEX REPLACE "[^A-Za-z0-9.+_-]" "_" _LIBRARY_NAME ${_LIBRARY_NAME})
                    set(_LIB_NAME CONAN_LIB::${package_name}_${_LIBRARY_NAME}${config_suffix})
                    if(NOT TARGET ${_LIB_NAME})
                        # Create a micro-target for each lib/a found

--- a/conans/client/generators/cmake_find_package_common.py
+++ b/conans/client/generators/cmake_find_package_common.py
@@ -135,6 +135,7 @@ class CMakeFindPackageCommonMacros:
                     list(APPEND _out_libraries ${CONAN_FOUND_LIBRARY})
                     if(NOT ${CMAKE_VERSION} VERSION_LESS "3.0")
                         # Create a micro-target for each lib/a found
+                        string(REGEX REPLACE "[^A-Za-z0-9.+_-]" "_" _LIBRARY_NAME ${_LIBRARY_NAME})
                         set(_LIB_NAME CONAN_LIB::${package_name}_${_LIBRARY_NAME}${build_type})
                         if(NOT TARGET ${_LIB_NAME})
                             # Create a micro-target for each lib/a found

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
@@ -35,8 +35,10 @@ def client_weird_lib_name():
                 self.copy("*.lib", dst="lib", keep_path=False)
                 self.copy("*.a", dst="lib", keep_path=False)
                 ext = "a" if platform.system() != "Windows" else "lib"
+                prefix = "lib" if platform.system() != "Windows" else ""
                 os.chdir(os.path.join(self.package_folder, "lib"))
-                os.rename("hello_0.1.{}".format(ext), "he!llo@0.1.{}".format(ext))
+                os.rename("{}hello_0.1.{}".format(prefix, ext),
+                          "{}he!llo@0.1.{}".format(prefix, ext))
 
             def package_info(self):
                 self.cpp_info.libs = ["he!llo@0.1"]

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
@@ -1,0 +1,111 @@
+import textwrap
+
+import pytest
+
+from conans.test.assets.cmake import gen_cmakelists
+from conans.test.assets.pkg_cmake import pkg_cmake
+from conans.test.assets.sources import gen_function_h, gen_function_cpp
+from conans.test.utils.tools import TestClient
+
+
+@pytest.fixture(scope="module")
+def client_weird_lib_name():
+    c = TestClient()
+    conanfile = textwrap.dedent("""
+        import os, platform
+        from conans import ConanFile
+        from conan.tools.cmake import CMake
+        from conan.tools.layout import cmake_layout
+
+        class Pkg(ConanFile):
+            exports_sources = "src/*"
+            settings = "os", "compiler", "arch", "build_type"
+            generators = "CMakeToolchain", "CMakeDeps"
+
+            def layout(self):
+                cmake_layout(self)
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure()
+                cmake.build()
+
+            def package(self):
+                self.copy("*.h", dst="include")
+                self.copy("*.lib", dst="lib", keep_path=False)
+                self.copy("*.a", dst="lib", keep_path=False)
+                ext = "a" if platform.system() != "Windows" else "lib"
+                os.chdir(os.path.join(self.package_folder, "lib"))
+                os.rename("hello_0.1.{}".format(ext), "he!llo@0.1.{}".format(ext))
+
+            def package_info(self):
+                self.cpp_info.libs = ["he!llo@0.1"]
+            """)
+
+    hdr = gen_function_h(name="hello")
+    src = gen_function_cpp(name="hello")
+    cmake = gen_cmakelists(libname="hello_0.1", libsources=["hello.cpp"])
+
+    c.save({"src/hello.h": hdr,
+            "src/hello.cpp": src,
+            "src/CMakeLists.txt": cmake,
+            "conanfile.py": conanfile})
+    c.run("create . hello/0.1@")
+    return c
+
+
+def test_cmakedeps(client_weird_lib_name):
+    c = client_weird_lib_name
+    c.save(pkg_cmake("chat", "0.1", requires=["hello/0.1"]), clean_first=True)
+    c.run("create . chat/0.1@")
+    assert "chat/0.1: Created package" in c.out
+
+
+# TODO: Remove in Conan 2.0
+def test_cmake_find_package(client_weird_lib_name):
+    c = client_weird_lib_name
+    files = pkg_cmake("chat", "0.1", requires=["hello/0.1"])
+    conanfile = textwrap.dedent("""
+        import os, platform
+        from conans import ConanFile, CMake
+
+        class Pkg(ConanFile):
+            exports_sources = "src/*"
+            settings = "os", "compiler", "arch", "build_type"
+            generators = "cmake_find_package"
+            requires = "hello/0.1"
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure(source_folder="src")
+                cmake.build()
+        """)
+    files["conanfile.py"] = conanfile
+    c.save(files, clean_first=True)
+    c.run("create . chat/0.1@")
+    assert "chat/0.1: Created package" in c.out
+
+
+# TODO: Remove in Conan 2.0
+def test_cmake_find_package_multi(client_weird_lib_name):
+    c = client_weird_lib_name
+    files = pkg_cmake("chat", "0.1", requires=["hello/0.1"])
+    conanfile = textwrap.dedent("""
+        import os, platform
+        from conans import ConanFile, CMake
+
+        class Pkg(ConanFile):
+            exports_sources = "src/*"
+            settings = "os", "compiler", "arch", "build_type"
+            generators = "cmake_find_package_multi"
+            requires = "hello/0.1"
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure(source_folder="src")
+                cmake.build()
+        """)
+    files["conanfile.py"] = conanfile
+    c.save(files, clean_first=True)
+    c.run("create . chat/0.1@")
+    assert "chat/0.1: Created package" in c.out

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_weird_library_names.py
@@ -56,6 +56,7 @@ def client_weird_lib_name():
     return c
 
 
+@pytest.mark.tool_cmake
 def test_cmakedeps(client_weird_lib_name):
     c = client_weird_lib_name
     c.save(pkg_cmake("chat", "0.1", requires=["hello/0.1"]), clean_first=True)
@@ -64,6 +65,7 @@ def test_cmakedeps(client_weird_lib_name):
 
 
 # TODO: Remove in Conan 2.0
+@pytest.mark.tool_cmake
 def test_cmake_find_package(client_weird_lib_name):
     c = client_weird_lib_name
     files = pkg_cmake("chat", "0.1", requires=["hello/0.1"])
@@ -89,6 +91,7 @@ def test_cmake_find_package(client_weird_lib_name):
 
 
 # TODO: Remove in Conan 2.0
+@pytest.mark.tool_cmake
 def test_cmake_find_package_multi(client_weird_lib_name):
     c = client_weird_lib_name
     files = pkg_cmake("chat", "0.1", requires=["hello/0.1"])

